File: MercuryClient/Web/Location/LocationResult.cs
/* 
 * Copyright (C) 2014 Alex Bikfalvi
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

using System;
using System.IO;
using System.Xml;
using Mercury.Xml;

namespace Mercury.Web.Location
{
	/// <summary>
	/// A class representing the result of a geo-location request.
	/// </summary>
	public sealed class LocationResult
	{
		/// <summary>
		/// Creates a new location result instance.
		/// </summary>
		private LocationResult()
		{
		}

		// Public methods.

		/// <summary>
		/// Gets the IP address.
		/// </summary>
		public string Address { get; private set; }
		/// <summary>
		/// Gets the country code.
		/// </summary>
		public string CountryCode { get; private set; }
		/// <summary>
		/// Gets the country name.
		/// </summary>
		public string CountryName { get; private set; }
		/// <summary>
		/// Gets the region code.
		/// </summary>
		public string RegionCode { get; private set; }
		/// <summary>
		/// Gets the region name.
		/// </summary>
		public string RegionName { get; private set; }
		/// <summary>
		/// Gets the city.
		/// </summary>
		public string City { get; private set; }
		/// <summary>
		/// Gets the ZIP code.
		/// </summary>
		public string ZipCode { get; private set; }
		/// <summary>
		/// Gets the latitude.
		/// </summary>
		public string Latitude { get; private set; }
		/// <summary>
		/// Gets the longitude.
		/// </summary>
		public string Longitude { get; private set; }
		/// <summary>
		/// Gets the metropolitan code.
		/// </summary>
		public string MetroCode { get; private set; }
		/// <summary>
		/// Gets the area code.
		/// </summary>
		public string AreaCode { get; private set; }

		// Public methods.

		/// <summary>
		/// Parses an XML string into a location result object.
		/// </summary>
		/// <param name="xml">The XML string.</param>
		/// <returns>The location result.</returns>
		public static LocationResult Parse(string xml)
		{
			// Create the XML document.
			XmlDocument document = new XmlDocument();

			// Create the location result object.
			LocationResult result = new LocationResult();

			using (StringReader reader = new StringReader(xml))
			{
				// Parse the string data into the XML document.
				document.Load(reader);
			}

			// Set the properties.
			result.Address = XmlExtensions.GetElementValue(document.DocumentElement, "Ip");
			result.CountryCode = XmlExtensions.GetElementValue(document.DocumentElement, "CountryCode");
			result.CountryName = XmlExtensions.GetElementValue(document.DocumentElement, "CountryName");
			result.RegionCode = XmlExtensions.GetElementValue(document.DocumentElement, "RegionCode");
			result.RegionName = XmlExtensions.GetElementValue(document.DocumentElement, "RegionName");
			result.City = XmlExtensions.GetElementValue(document.DocumentElement, "City");
			result.ZipCode = XmlExtensions.GetElementValue(document.DocumentElement, "ZipCode");
			result.Latitude = XmlExtensions.GetElementValue(document.DocumentElement, "Latitude");
			result.Longitude = XmlExtensions.GetElementValue(document.DocumentElement, "Longitude");
			result.MetroCode = XmlExtensions.GetElementValue(document.DocumentElement, "MetroCode");
			result.AreaCode = XmlExtensions.GetElementValue(document.DocumentElement, "AreaCode");

			// Return the result.
			return result;
		}

		// Private methods.

		private static XmlElement GetElement(XmlElement element, string name)
		{
			// Get the child elements.
			XmlNodeList children = element.GetElementsByTagName(name);
			// If the list is not empty, return the first element.
			return children.Count > 0 ? children[0] as XmlElement : null;
		}
	}
}

File: Eclipse/MercuryClient/src/com/bikfalvi/java/web/location/LocationResult.java
/** 
 * Copyright (C) 2013 Alex Bikfalvi
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

package com.bikfalvi.java.web.location;

import java.io.IOException;
import java.io.InputStream;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.xml.sax.SAXException;

import com.bikfalvi.java.xml.XmlExtensions;

/**
 * A class representing the result of a geo-location request.
 * @author Alex Bikfalvi
 *
 */
public final class LocationResult
{
	private String address;
	private String countryCode;
	private String countryName;
	private String regionCode;
	private String regionName;
	private String city;
	private String zipCode;
	private String latitude;
	private String longitude;
	private String metroCode;
	private String areaCode;
	
	/**
	 * Creates a new location result instance.
	 */
	private LocationResult()
	{
	}
	
	/**
	 * Parses an XML String into a location result object.
	 * @param stream The XML string.
	 * @return The location result.
	 * @throws ParserConfigurationException 
	 * @throws IOException 
	 * @throws SAXException 
	 */
	public static LocationResult parse(InputStream stream) throws ParserConfigurationException, SAXException, IOException
	{
		// Create the document builder factory.
		DocumentBuilderFactory domFactory = DocumentBuilderFactory.newInstance();
		// Create the document builder.
		DocumentBuilder domBuilder = domFactory.newDocumentBuilder();
		
		// Parse the XML document from the input string.
		Document document = domBuilder.parse(stream);

		// Create the location result object.
		LocationResult result = new LocationResult();

		// Set the properties.
		result.address = XmlExtensions.getElementValue(document.getDocumentElement(), "Ip");
		result.countryCode = XmlExtensions.getElementValue(document.getDocumentElement(), "CountryCode");
		result.countryName = XmlExtensions.getElementValue(document.getDocumentElement(), "CountryName");
		result.regionCode = XmlExtensions.getElementValue(document.getDocumentElement(), "RegionCode");
		result.regionName = XmlExtensions.getElementValue(document.getDocumentElement(), "RegionName");
		result.city = XmlExtensions.getElementValue(document.getDocumentElement(), "City");
		result.zipCode = XmlExtensions.getElementValue(document.getDocumentElement(), "ZipCode");
		result.latitude = XmlExtensions.getElementValue(document.getDocumentElement(), "Latitude");
		result.longitude = XmlExtensions.getElementValue(document.getDocumentElement(), "Longitude");
		result.metroCode = XmlExtensions.getElementValue(document.getDocumentElement(), "MetroCode");
		result.areaCode = XmlExtensions.getElementValue(document.getDocumentElement(), "AreaCode");

		// Return the result.
		return result;
	}

	/**
	 * Gets the IP address.
	 * @return The IP address.
	 */
	public String getAddress() {
		return address;
	}

	/**
	 * Gets the country code.
	 * @return The country code.
	 */
	public String getCountryCode() {
		return countryCode;
	}

	/**
	 * Gets the country name.
	 * @return The country name.
	 */
	public String getCountryName() {
		return countryName;
	}

	/**
	 * Gets the region code.
	 * @return The region code.
	 */
	public String getRegionCode() {
		return regionCode;
	}

	/**
	 * Gets the region name.
	 * @return The region name.
	 */
	public String getRegionName() {
		return regionName;
	}

	/**
	 * Gets the city.
	 * @return The city.
	 */
	public String getCity() {
		return city;
	}

	/**
	 * Gets the ZIP code.
	 * @return The ZIP code.
	 */
	public String getZipCode() {
		return zipCode;
	}

	/**
	 * Gets the latitude.
	 * @return The latitude.
	 */
	public String getLatitude() {
		return latitude;
	}

	/**
	 * Gets the longitude.
	 * @return The longitude.
	 */
	public String getLongitude() {
		return longitude;
	}

	/**
	 * Gets the metro code.
	 * @return The metro code.
	 */
	public String getMetroCode() {
		return metroCode;
	}

	/**
	 * Gets the area code.
	 * @return The area code.
	 */
	public String getAreaCode() {
		return areaCode;
	}
}
